r2/r2/lib/media.py

```diff
@@ -205,6 +452,83 @@ def optimize_jpeg(filename):
     with open(os.path.devnull, 'w') as devnull:
         subprocess.check_call(("/usr/bin/jpegoptim", filename), stdout=devnull)
 
+class _SVGSoup(BeautifulSoup.BeautifulStoneSoup):
+    # Fixes unexpected conversion from `<g><g></g></g>` into `<g></g><g></g>`
+    nests = ('defs', 'g', 'marker', 'mask', 'pattern', 'svg', 'symbol', 'use')
+    NESTABLE_TAGS = BeautifulSoup.buildTagMap([], nests)
+
+def _sanitize_svg(image):
+    elements = [x.lower() for x in _SAFE_SVG_ELEMENTS]
+    attributes = [x.lower() for x in _SAFE_SVG_ATTRIBUTES]
+
+    # BS3 gets confused with some self-closing tags. We can safely 
+    # convert them with regex and afterward, we strip out unwanted <!> 
+    # and <?> tags.
+    #
+    # Keep in mind that `[^>]` goes in place of `.`, because it matches
+    # newline chars as well. One could also flag re.DOTALL, but we need
+    # to also not match `>`s.
+    image = re.sub(r'<([^>\s]+)\s*([^>]*)/>', r'<\1 \2></\1>',
+                   re.sub(r'<![^>]*>|<\?(?!xml)[^>]*>', r'', image))
+    soup = _SVGSoup(image, 
+                    convertEntities=BeautifulSoup.BeautifulStoneSoup.XML_ENTITIES)
+
+    inspect_element = soup.findAll(True)
+
+    for elem in inspect_element:
+        if elem.name not in elements: # Whitelist elements for security
+            if elem.contents and elem.parent:
+                parent = elem.parent
+                index = parent.contents.index(elem)
+
+                # Put all children in place of disallowed element
+                for child in elem.findAll(True, recursive=False):
+                    parent.insert(index + 1, child)
+                    inspect_element.append(child)
+                    index += 1
+            
+            elem.decompose()
+            continue
+
+        if elem.attrs:
+            to_del = set()
+            to_save = dict()
+
+            for attribute_set in elem.attrs:
+                attr = attribute_set[0]
+                value = attribute_set[1]
+                to_del.update({attr})
+
+                # When safe:
+                # Expression 1 == True
+                # Expression 2 == None
+                # Expression 3 == False
+
+                if (attr in attributes and
+                    re.search(r'url\((?!#)', elem[attr]) == None and
+                        (attr == "xlink:href" and not 
+                         elem[attr].startswith('#')) == False):
+
+                    # BS3 lowercases attribute names - this can be bad
+                    # Using the two arrays we made, we can revert this
+                    # (This is also why we delete every attribute first)
+                    new_attr = _SAFE_SVG_ATTRIBUTES[attributes.index(attr)]
+                    to_save[new_attr] = elem[attr]
+
+            for key in to_del:
+                del elem[key]
```

Comment:
It'd be better to create a version of the tree with only known-clean attrs and elems, rather than mutate the existing tree. That way you're not vulnerable to weird issues if whatever you're using to walk the tree misses something security-relevant (a doctype decl, a PI, a comment, etc.)